docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,936 @@
+
+```
+COM instance
+0x10000  0x20000
+
+VTable for IDemoStoreType
+0x20000  <Address of QueryInterface>
+0x20008  <Address of AddRef>
+0x20010  <Address of Release>
+0x20018  <Address of StoreString>
+```
+
+Having the function pointer then allows us to dispatch to that member function on that object by passing the object instance as the fist parameter &ndash; that should look familiar based on the function definitions when implementing the Managed Object Wrapper.
+
+Once the `CreateObject()` method is implemented the `ComWrappers` subclass will be able to produce Native Object Wrappers for COM instances that implement both `IDemoGetType` and `IDemoStoreType`.
+
+```csharp
+IntPtr iunk = ...; // Get a COM instance from native code.
+object rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);
+```
+
+### Step 4 &ndash; Native Object Wrapper lifetime details
+
+The `ComputeVtables()` and `CreateObject()` implementations discussed some wrapper lifetime details but further details need discussion. This can be a short step but can also significantly increase the complexity of the `ComWrappers` design.
+
+Unlike the Managed Object Wrapper, which is controlled by calls to its `AddRef()`/`Release()`, the lifetime of a Native Object Wrapper is non-deterministically handled by the GC. The question here is, when does the Native Object Wrapper call `Release()` on the `IntPtr` that represents the COM instance? There are two general buckets here.
+
+1) The Native Object Wrapper's Finalizer should be responsible for calling the COM instance's `Release()`. This is the only time when it is safe to call this method since at this point it has been correctly determined by the GC there are no other references to the Native Object Wrapper in the .NET runtime. There can be complexity here if one is properly supporting COM Apartments &ndash; see the "Additional considerations" section.
+
+2) Have the returned Native Object Wrapper implement `IDisposable` and call `Release()` in `Dispose()`. This pattern should only be supported if during the `CreateObject()` call the `CreateObjectFlags.UniqueInstance` flag was passed in. **Note** if this requirement is not followed it is possible for disposed Native Object Wrappers to be reused after being disposed.
+
+## Using the `ComWrappers` subclass
+
+We now have a `ComWrappers` subclass that can be tested. In order to avoid creating a native library that returns a COM instance that implements `IDemoGetType` and `IDemoStoreType` we are going to use the Managed Object Wrapper and treat it as a COM instance &ndash; this must be possible in order to pass it COM anyways.
```

Comment:
```suggestion
You now have a `ComWrappers` subclass that can be tested. To avoid creating a native library that returns a COM instance that implements `IDemoGetType` and `IDemoStoreType`, you'll use the Managed Object Wrapper and treat it as a COM instance &ndash; this must be possible in order to pass it COM anyways.
```